pubsublite/internal/test/verifier.go

```diff
@@ -0,0 +1,225 @@
+	return &RPCVerifier{
+		t:        t,
+		rpcs:     list.New(),
+		numCalls: -1,
+	}
+}
+
+// Push appends a new {request, response, error} tuple.
+func (v *RPCVerifier) Push(wantRequest interface{}, retResponse interface{}, retErr error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	v.rpcs.PushBack(&rpcMetadata{
+		wantRequest: wantRequest,
+		retResponse: retResponse,
+		retErr:      retErr,
+	})
+}
+
+// PushWithBlock is like Push, but returns a channel that the test should close
+// when it would like the response to be sent to the client. This is useful for
+// synchronizing with work that needs to be done on the client.
+func (v *RPCVerifier) PushWithBlock(wantRequest interface{}, retResponse interface{}, retErr error) chan struct{} {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	block := make(chan struct{})
+	v.rpcs.PushBack(&rpcMetadata{
+		wantRequest:   wantRequest,
+		retResponse:   retResponse,
+		retErr:        retErr,
+		blockResponse: block,
+	})
+	return block
+}
+
+// Pop validates the received request with the next {request, response, error}
+// tuple.
+func (v *RPCVerifier) Pop(gotRequest interface{}) (interface{}, error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	v.numCalls++
+	elem := v.rpcs.Front()
+	if elem == nil {
+		v.t.Errorf("call(%d): unexpected request:\n%v", v.numCalls, gotRequest)
+		return nil, status.Error(codes.FailedPrecondition, "mockserver: got unexpected request")
+	}
+
+	rpc, _ := elem.Value.(*rpcMetadata)
+	v.rpcs.Remove(elem)
+
+	if !testutil.Equal(gotRequest, rpc.wantRequest) {
+		v.t.Errorf("call(%d): got request: %v\nwant request: %v", v.numCalls, gotRequest, rpc.wantRequest)
+	}
+	if err := rpc.wait(); err != nil {
+		return nil, err
+	}
+	return rpc.retResponse, rpc.retErr
+}
+
+// TryPop should be used only for streams. It checks whether the request in the
+// next tuple is nil, in which case the response or error should be returned to
+// the client without waiting for a request. Useful for streams where the server
+// continuously sends data (e.g. subscribe stream).
+func (v *RPCVerifier) TryPop() (bool, interface{}, error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	elem := v.rpcs.Front()
+	if elem == nil {
+		return false, nil, nil
+	}
+
+	rpc, _ := elem.Value.(*rpcMetadata)
+	if rpc.wantRequest != nil {
+		return false, nil, nil
+	}
+
+	v.rpcs.Remove(elem)
+	if err := rpc.wait(); err != nil {
+		return true, nil, err
+	}
+	return true, rpc.retResponse, rpc.retErr
+}
+
+// Flush logs an error for any remaining {request, response, error} tuples, in
+// case the client terminated early.
+func (v *RPCVerifier) Flush() {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	for elem := v.rpcs.Front(); elem != nil; elem = elem.Next() {
+		v.numCalls++
+		rpc, _ := elem.Value.(*rpcMetadata)
+		v.t.Errorf("call(%d): did not receive expected request:\n%v", v.numCalls, rpc.wantRequest)
+	}
+	v.rpcs.Init()
+}
+
+// streamVerifiers stores a stack of verifiers for unique stream connections.
```

Comment:
Same comment as earlier, is this supposed be a queue? If not, could also change the `PushBack` calls to `PushFront`